shpc/main/modules/__init__.py

```diff
@@ -359,3 +363,45 @@ def install(self, name, tag=None, **kwargs):
             name = "%s:%s" %(name, tag.name)
         logger.info("Module %s was created." % name)
         return container_path
+
+    # Module software can choose how to handle each of these cases
+    def _no_default_version(self, version_file, tag):
+        return
+
+    def _module_sys_default_version(self, version_file, tag):
+        return
+
+    def _set_default_version(self, version_file, tag):
+        """
+        Set the default version to the given tag
+        """
+        template = self._load_template("default_version")
+        utils.write_file(version_file, template.render(version=tag))
+
+    def write_version_file(self, uri, latest_tag_installed=None):
+        """
+        Create the .version file, if there is a template for it.
+        """
+        version_dir = os.path.join(self.settings.module_base, uri)
+        version_file = os.path.join(version_dir, ".version")
+
+        # No default versions
+        if self.settings.default_version in [False, None]:
+            self._no_default_version(version_file, latest_tag_installed)
+
+        # allow the module software to control versions
+        elif self.settings.default_version in [True, "module_sys"]:
+            self._module_sys_default_version(version_file, latest_tag_installed)
+
+        # First or last installed
+        else:
+            # The versions we actually have
+            found = [x for x in os.listdir(version_dir) if x != ".version"]
+            if len(found) == 1:
+                tag = found[0]
+            else:
+                selector = min if self.settings.default_version == "first_installed" else max
+                tag = selector(found, key=lambda x: utils.creation_date(os.path.join(version_dir, x)))
+
+            # Write the .version file
+            self._set_default_version(version_file, tag)
```

Comment:
Just a suggestion to warrant the extra else block not needed!
```suggestion
        if self.settings.default_version in [False, None]:
            return self._no_default_version(version_file, latest_tag_installed)

        # allow the module software to control versions
        elif self.settings.default_version in [True, "module_sys"]:
            return self._module_sys_default_version(version_file, latest_tag_installed)

        # First or last installed
        # The versions we actually have
        found = [x for x in os.listdir(version_dir) if x != ".version"]
        if len(found) == 1:
            tag = found[0]
        else:
            selector = min if self.settings.default_version == "first_installed" else max
            tag = selector(found, key=lambda x: utils.creation_date(os.path.join(version_dir, x)))

        # Write the .version file
        return self._set_default_version(version_file, tag)
```
Technically they all return None, so the return at the bottom is to be consistent, and the others are to exit early. And maybe one day we will have reason to know some result (e.g., True for changed False otherwise) but we don't see that yet!

I've also walked through the logic of different scenarios, and I don't see any holes! E.g., if we are installing the first time, the module_dir will always create that first version, and then the tag is provided OR we derive at found[0]. There can never be a case where len(found) == 0 because the cleanup function should handle removing empty directories, and the check in uninstall doesn't call the function if the module directory doesn't exist.

Reply:
Also I like the selector bit you added and using min and max,  it's pretty neat :)